### ISSUE TITLE
Filter while searching by find files and live grep

### ISSUE THREAD

**Issue by ecosse3** [nvim-telescope/telescope.nvim#598]
It would be nice to have a feature to _filter as you type_ in project files and live grep.

**How does it work?**
Personally, I've used just two filters that are implemented in [fzf-preview.vim](https://github.com/yuki-yano/fzf-preview.vim) extensions. I think there could be (as in fzf-preview) more useful filters.

The first one is filtering via filename/filetype. All you need is to type **'something** in filename/filetype (single quote at beginning) and then search results will be filtered:

It works like you can type multiple words and just this one with a single quote at the beginning will be filtering in filename and other words as usual. You can also connect multiple filters together (next example below).

![image](https://user-images.githubusercontent.com/17691791/109865949-fdb4f200-7c64-11eb-8552-3bd5e8c51b65.png)
![image](https://user-images.githubusercontent.com/17691791/109865974-060d2d00-7c65-11eb-9d4a-f2bbf5855b57.png)

Another filter I use a lot is **not** (exclamation mark). It allows you to search for project files or grep inside but not with what you provided with _exclamation mark_. So if you type for example "useEffect !from 'tsx" it will show you all file lines with "useEffect" in filetype "tsx" but not lines with "from", like below:

![image](https://user-images.githubusercontent.com/17691791/109866632-c7c43d80-7c65-11eb-92ae-7786be8ebe09.png)

Without "!from" it will show results more results:
![image](https://user-images.githubusercontent.com/17691791/109866690-dd396780-7c65-11eb-8d86-986140003cd8.png)

I would extremely love this feature in telescope.nvim and I think it's very useful. Thanks a lot.

**Comment by Conni2461** [comment]
We have talked about this a couple of times. Because we can replace sorter something like this is possible. For example this: https://github.com/nvim-telescope/telescope.nvim/commit/db7615578b43586ef6526c56ba397f86b13fbed9 just introduced filtering for `lsp_*_symbols`. Like `:function:` will return all functions when doing `lsp_document_symbols` and than you can filter over those. I am not sure tho if we should build the whole fzf syntax our self. Like fzf syntax has a lot https://github.com/junegunn/fzf#search-syntax.

We had a discussion over here: https://github.com/nvim-telescope/telescope.nvim/issues/550 and i suggested we could do something similar to `fzy-native` (implement fzf in c and use lua ffi). See https://github.com/nvim-telescope/telescope.nvim/issues/550#issuecomment-785268386
Its just not that easy because `fzf` isn't a lib and written in go. So the probably more practical approach is build skim as dylib and use write ffi for that. I looked into it a little bit but i am currently kinda too busy for such a big task.

tldr: Its likely that we will have a something like this in the future just a lot of work.

**Comment by TC72** [comment]
Just to add to this, the best way to describe ' is it turns off fuzzy searching so you'll get an exact match of what you type.
Fzf has lots of these filters and it would be great to have them in Telescope.

**Comment by carloscoves** [comment]
I only use
> `'wild`	exact-match (quoted)

Is there any way we can do exact-match (no fuzzy filtering) right now?

**Comment by Conni2461** [comment]
I am working on a `telescope-fzf-native.nvim` and i am already pretty far. See for the progress:
- https://github.com/nvim-telescope/telescope.nvim/issues/748#issuecomment-817287545
- https://github.com/nvim-telescope/telescope.nvim/issues/748#issuecomment-817341079
- https://github.com/nvim-telescope/telescope.nvim/issues/748#issuecomment-817356783

Right now it should be done like maybe end of week. But its `c` so things can go wrong pretty fast :rofl:

**Comment by Conni2461** [comment]
Status update abou `telescope-fzf-native.nvim` progress for everyone subbed to this thread:
![image](https://user-images.githubusercontent.com/15233006/114577581-68057d00-9c7c-11eb-965c-2bfe4a4af175.png)

Missing:
- invert
- and probably more, i am not thinking about right now :rofl:

**Comment by Conni2461** [comment]
https://github.com/nvim-telescope/telescope-fzf-native.nvim

Read the friendly readme ;)

Also highlighting for `prefix_match` does not work because how telescope handles things. But i didn't want that you have to wait until eternity :laughing: i will fix that at some point

CC @phaazon

**Comment by Conni2461** [comment]
forgot to mention requires up to date telescope :)

**Comment by ecosse3** [comment]
Thank you for this! I really appreciate it. For me, that's an amazing feature in telescope now. Will be waiting for new features added to fzf-native.
